resource_providers/src/http.rs

```diff
@@ -0,0 +1,176 @@
+    ClientError { cause: String },
+    #[error("guest produced a malformed request")]
+    BadRequest,
+    /// This should never happen. If we cannot find a client the guest is
+    /// either somehow intentionally requesting access to resources it should not access
+    /// Or, more likely we have a bug tracking the lifetime of http client resources
+    #[error("guest tried to use a client which does not exist")]
+    HostResourceNotFound,
+}
+
+/// Abstract service type defining an HttpClient
+/// which can be provided by specific concrete implementations
+/// for example we have a `HyperHostClient`
+pub type HostClient = BoxService<ClientRequest, ClientResult, HttpClientError>;
+
+/// Abstract service type for a thing which produces new HostClients
+/// for example see `providers::hyper::new_client_maker`
+pub type HostClientMaker = BoxService<(), HostClient, ClientMakeError>;
+
+/// Manages the association between guest wasm http clients and their host implementation instances.
+pub struct HostClientResource {
+    pub resource_id_provider: HostResourceIdProvider,
+    pub client_maker: BoxService<(), HostClient, ClientMakeError>,
+    pub clients: HashMap<u32, HostClient>,
+}
+
+impl HostClientResource {
+    pub fn new(
+        client_maker: HostClientMaker,
+        resource_id_provider: HostResourceIdProvider,
+    ) -> Self {
+        Self {
+            resource_id_provider,
+            client_maker,
+            clients: Default::default(),
+        }
+    }
+}
+
+#[async_trait]
+impl HostClientInterface for HostClientResource {
+    /// Creates a new HttpClient Resource and returns the resource to the guest.
+    ///
+    /// We return a resource id the guest will use to request access to the resource
+    /// internally, we create a mapping of <ID, Client> so we can lookup the correct
+    /// client to make the request.
+    ///
+    /// Note, in the future this will allow us to have precise control over
+    /// the behavior of the host client. Rate limiting, logging, etc, request tagging, etc :D
+    async fn new(&mut self) -> anyhow::Result<Resource<Client>> {
+        let rdy_client = self.resource_id_provider.ready().await?;
+        let new_id = rdy_client.call(()).await?;
+
+        let rdy_client = self.client_maker.ready().await?;
+        let new_client = rdy_client.call(()).await?;
+
+        if let Some(_) = self.clients.insert(new_id, new_client) {
+                      // print error. This is indicative of a bug in the upstream id provider client
+            return Err(ClientMakeError::BadResourceId);
+        }
+
+        Ok(Resource::new_own(new_id))
+    }
+
+    /// Attempts to make an HttpRequest using some resource
+    async fn send(
+        &mut self,
+        guest_self: Resource<Client>,
+        req: ClientRequest,
+    ) -> anyhow::Result<ClientResult> {
+        let id = guest_self.rep();
+        let client = self.clients.get_mut(&id).ok_or(HttpClientError::HostResourceNotFound)?;
```

Comment:
Why aren't we `panic`'ing on `HostResourceNotFound`? Its comment states that it should never happen

Reply:
I'm not entirely sure how we should handle this case tbh - lets panic. Good call